AI #1785: Add Invoice Detail and Billing Period datasets (#1900)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/invoice_detail/columns/billedcost.md

  * The sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) when [ChargeCategory](#datasets.invoicedetail.chargecategory) = "Tax" or [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open".
  * Otherwise, the absolute difference between the sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) and the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST NOT exceed the **greater** of:

>     * 100 units of the smallest subunit of the [BillingCurrency](#datasets.invoicedetail.billingcurrency) (e.g., 1.00 for USD, 100 for JPY).

Reviewer: @mike-finopsorg, @shawnalpay, These two statements don't contain a BCP14 keyword, and I am unclear how to incorporate this into the model.

Author: @jpradocueva Do you have a suggestion?  I can fold them into one long sentence, but it will be harder to read.

Reviewer 2: ```
Otherwise, the absolute difference between the sum of BilledCost for a given InvoiceId and InvoiceIssuerName and the sum of CostAndUsage.BilledCost for the same CostAndUsage.InvoiceId and CostAndUsage.InvoiceIssuerName MUST NOT exceed the greater of:
* 100 units of the smallest subunit of the BillingCurrency (e.g., 1.00 for USD, 100 for JPY).
* The square root of the count of CostAndUsage rows multiplied by 0.5 times the smallest subunit of the BillingCurrency.

```

Otherwise here means (when ChargeCategory != "Tax" and InvoiceIssueStatus is not "Open".) We should be explicit.

The two lines are not "extra normative lines" but part of the "exceed the greater of" Wondering if we need to define the method outside the normative text and then have the normative text reference the definition?

Author: @mike-finopsorg Happy to consider a suggestion if you have it!

>     * The square root of the count of [CostAndUsage](#datasets.costandusage) rows multiplied by 0.5 times the smallest subunit of the [BillingCurrency](#datasets.invoicedetail.billingcurrency).

Reviewer 2: ```suggestion
* When [ChargeCategory](#datasets.invoicedetail.chargecategory) is not "Tax" and [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is not "Open", the sum of BilledCost aggregated by [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MUST NOT differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) aggregated by [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) by more than `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)` as defined in [Tolerance Formula](#datasets.invoicedetail.billedcost.implementationguidance.toleranceformula).
* When [ChargeCategory](#datasets.invoicedetail.chargecategory) is "Tax" or [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open", the sum of BilledCost aggregated by [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) aggregated by [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername).
```

>     * The square root of the count of [CostAndUsage](#datasets.costandusage) rows multiplied by 0.5 times the smallest subunit of the [BillingCurrency](#datasets.invoicedetail.billingcurrency).

Reviewer 2: ```suggestion
* When comparing BilledCost aggregated by [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) with [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) aggregated by [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername), BilledCost MUST adhere to the following requirements:
  * When [ChargeCategory](#datasets.invoicedetail.chargecategory) is not "Tax" and [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is not "Open", the sum of BilledCost MUST NOT differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) by more than `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)` as defined in [Tolerance Formula](#datasets.invoicedetail.billedcost.implementationguidance.toleranceformula).
  * When [ChargeCategory](#datasets.invoicedetail.chargecategory) is "Tax" or [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open", the sum of BilledCost MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost).
```

  * Otherwise, the absolute difference between the sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) and the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceDetailId](#datasets.costandusage.invoicedetailid), [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid), and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST NOT exceed the **greater** of:
    * 100 units of the smallest subunit of the [BillingCurrency](#datasets.invoicedetail.billingcurrency) (e.g., 1.00 for USD, 100 for JPY).

>     * The square root of the count of [CostAndUsage](#datasets.costandusage) rows multiplied by 0.5 times the smallest subunit of the [BillingCurrency](#datasets.invoicedetail.billingcurrency).

Reviewer 2: ```suggestion
* When comparing BilledCost aggregated by [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) with [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) aggregated by [CostAndUsage.InvoiceDetailId](#datasets.costandusage.invoicedetailid), [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid), and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername), BilledCost MUST adhere to the following requirements:
  * When [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is not "Open", the sum of BilledCost MUST NOT differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) by more than `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)` as defined in [Tolerance Formula](#datasets.invoicedetail.billedcost.implementationguidance.toleranceformula).
  * When [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open", the sum of BilledCost MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost).
```

## Implementation Guidance

### Handling Rounding Discrepancies

> When validating InvoiceDetail.BilledCost against [CostAndUsage.BilledCost](#datasets.costandusage.billedcost), exact matches are not expected due to precision differences (e.g., 6 decimals in CostAndUsage vs. 2 decimals in InvoiceDetail). The requirement allows for a maximum rounding error based on the statistical probability of rounding variance, which grows with the square root of the row count.

Reviewer 3: Thanks for adding this, overall it is a well reasoned approach

Author: Awesome, glad you like it!

> When validating InvoiceDetail.BilledCost against [CostAndUsage.BilledCost](#datasets.costandusage.billedcost), exact matches are not expected due to precision differences (e.g., 6 decimals in CostAndUsage vs. 2 decimals in InvoiceDetail). The requirement allows for a maximum rounding error based on the statistical probability of rounding variance, which grows with the square root of the row count.

Reviewer 3: 6- 9 decimals in CostandUsage

Author: Yes.  Precision could be whatever the data generator decides (we don't yet specify scale differences in FOCUS), but I used 6 as an example.

### Tolerance Formula

Tolerance = `MAX(100 * Subunit, SQRT(Rows) * 0.5 * Subunit)`

* Rows: The count of related [CostAndUsage](#datasets.costandusage) rows.
* Subunit: The smallest subunit of [BillingCurrency](#datasets.invoicedetail.billingcurrency) (e.g., 0.01 for USD, 1 for JPY).

> * 100 * Subunit: Represents a universal floor of 100 "ticks" of precision to prevent false positives on small files, regardless of currency scale.

Reviewer 2: ```suggestion
### Tolerance Formula

The tolerance used when comparing aggregated BilledCost values is defined as:

Tolerance = `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)`

Where:

* **Rows** — The number of [CostAndUsage](#datasets.costandusage) rows included in the aggregation for the relevant [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername).
* **Subunit** — The numeric value of the smallest subunit of [BillingCurrency](#datasets.invoicedetail.billingcurrency) (for example, `0.01` for USD or `1` for JPY).

The tolerance is the **greater** of the following values:

* **100 × Subunit** — Establishes a fixed minimum tolerance equal to 100 units of the smallest currency subunit.
* **(SQRT(Rows) × 0.5) × Subunit** — Provides a tolerance that increases with the square root of the number of rows to account for rounding accumulation in larger datasets.

```

* Limit Calculation:
  * Statistical Limit: `SQRT(5) * 0.5 * 0.01` = &dollar;0.011.
  * Floor Limit: `100 * 0.01` = &dollar;1.00.
  * Effective Tolerance: **&dollar;1.00** (Greater of &dollar;0.011 and &dollar;1.00).
* Result: **Pass** (Difference &dollar;0.02 < Tolerance &dollar;1.00).

### Scenario 2: Small Invoice (Fail)

* Data: A small invoice with 5 line items where a &dollar;5.00 charge is missing from CostAndUsage.
* Values: CostAndUsage sums to &dollar;10.00. InvoiceDetail sums to &dollar;15.00. Difference is **&dollar;5.00**.
* Limit Calculation:
  * Effective Tolerance: **&dollar;1.00**.
* Result: **Fail** (Difference &dollar;5.00 > Tolerance &dollar;1.00).

### Scenario 3: Large Invoice (Pass)

* Data: An enterprise invoice with 1,000,000 line items.
* Values: CostAndUsage sums to &dollar;5,000,000.00. InvoiceDetail sums to &dollar;5,000,004.50. Difference is **&dollar;4.50**.
* Limit Calculation:
  * Statistical Limit: `SQRT(1,000,000) * 0.5 * 0.01` = &dollar;5.00.
  * Floor Limit: `100 * 0.01` = &dollar;1.00.
  * Effective Tolerance: **&dollar;5.00** (Greater of &dollar;5.00 and &dollar;1.00).
* Result: **Pass** (Difference &dollar;4.50 < Tolerance &dollar;5.00).

> ### Scenario 4: Large Invoice (Fail)

Reviewer 3: I am trying to pressure this with a large invoice and or/CUR data with a huge number of rows. will get back by 3/3.

Author: Discussed this in Mar 3 TF-1.  @pydi-aws will work with his team on this in the next couple of days.  Targeting giving final feedback by end of week.

Reviewer 3: I am still trying to get a query setup to capture the data - will close this by Tuesday next week.

Author: Thank you @pydi-aws!

Author: Discussed in Mar 10 TF-1.  @pydi-aws will look to resolve in the next 24 hours.  We will still proceed with targeting Member approval in Thu Mar 12 Members.

Reviewer 3: I pressure-tested the tolerance formula against a AWS payer account with ~1.97 billion CUR rows across 30 invoices covering subscription, usage, and marketplace charge type.

Result: all 30 invoices pass.

The SQRT-based formula with the $1.00 floor works well with the floor catches all subscription invoices (low row counts, hundreds of rows), while the SQRT scaling provides appropriate room for high-volume usage invoices. Also, validated that the formula correctly fails when real data gaps exist confirming it distinguishes legitimate rounding drift from actual data issues. We should be good to close this thread.

Author: Excellent!  I'm super glad to hear that.  Resolving this thread -- thank you @pydi-aws!

# Billed Cost

> TODO: Apply the results of the work from PR #1678.

Reviewer 2: Just calling out this TODO so its visible on the PR

Author: Thankfully PR #1678 looks on target for getting approved this coming Thu.  I'll fill it in then.

Author: This one is now resolved!

~~The [*billed cost*](#glossary:billed-cost) represents a [*charge*](#glossary:charge) serving as the basis for invoicing, inclusive of the impacts of all reduced rates and discounts while excluding the [*amortization*](#glossary:amortization) of relevant purchases (one-time or recurring) paid to cover future eligible *charges*. This cost is denominated in the [Billing Currency](#datasets.costandusage.billingcurrency). The Billed Cost is commonly used to perform FinOps capabilities that require cash-basis accounting such as cost allocation, budgeting, and invoice reconciliation.~~

## Requirements

> TODO: Apply the results of the work from FR #982.

Reviewer 2: Just calling out this TODO so its visible on the PR

Author: Yes.  That body of work from #982 may not get hammered out until near the end of the 1.4 dev cycle, but we need to get this PR over the line so we can move forward on correction handling.

I can make a note of reconciliation during consistency review, and we can move forward with the "old" BilledCost normative requirements here.

Author: This verbiage will be dependent on us approving the TF-3 PR https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1966.  I have made a note to address this in 1.4 consistency review via AI #1878.  Can we resolve, or would you prefer to hold open?

Author: Per my note from last week, I have captured this on #1878.  Resolving this thread.

BilledCost adheres to the following requirements:

* BilledCost MUST be present in an Invoice Detail [*FOCUS dataset*](#glossary:FOCUS-dataset).
* BilledCost MUST be of type Decimal.
* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be 0 for *charges* where payments are received by a third party (e.g., marketplace transactions).
* BilledCost MUST be denominated in the BillingCurrency.
* The sum of the BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same ID generated by the [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername).
* The sum of the BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid) MUST match the sum of the payable amount provided in the corresponding invoice.

## Column ID

BilledCost

## Display Name

Billed Cost

## Description

> TODO: Apply the results of the work from PR #1678.

Reviewer 2: Just calling out this TODO so its visible on the PR

Author: Thankfully PR https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1678 looks on target for getting approved this coming Thu. I'll fill it in then.

Author: Addressed!  Resolving.

# Billed Cost

Billed Cost represents the cost of a [*charge*](#glossary:charge) as invoiced by the [invoice issuer](#datasets.invoicedetail.invoiceissuername) in a given [*billing period*](#glossary:billing-period).

For all *charges*, Billed Cost reflects all applicable pricing adjustments (e.g., reduced pricing from [*negotiated discounts*](#glossary:negotiated-discount) or [*commitment discounts*](#glossary:commitment-discount)). For purchase *charges*, Billed Cost includes any portion invoiced in the given *billing period*. For usage *charges*, Billed Cost excludes any portion covered by related purchase *charges* (e.g., *commitments*, pre-payments, or marketplace purchases), regardless of when those related charges are invoiced.

Billed Cost is denominated in the [Billing Currency](#datasets.invoicedetail.billingcurrency). Billed Cost is commonly used to support FinOps activities, including invoice reconciliation, [*cash-flow-based*](#glossary:cash-based-accounting) forecasting, budgeting, and cost allocation.

## Requirements

> BilledCost adheres to the following requirements:

Reviewer: ```suggestion
BilledCost MUST adhere to the following requirements:
```

* BilledCost MUST be of type Decimal.
* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be 0 for *charges* where payments are received by a third party (e.g., marketplace transactions).
* BilledCost MUST be denominated in the BillingCurrency.
* The sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MUST match the payable amount provided in the corresponding entries on the issued invoice when [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Issued".

> * When comparing BilledCost and [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername), BilledCost adheres to the following requirements:

Reviewer: ```suggestion
* When comparing BilledCost and [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername), BilledCost MUST adhere to the following requirements:
```

  * The sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) when [ChargeCategory](#datasets.invoicedetail.chargecategory) = "Tax" or [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open".
  * Otherwise, the sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MUST match the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername).

> * When comparing BilledCost and [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername), BilledCost adheres to the following requirements:

Reviewer: ```suggestion
* When comparing BilledCost and [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid) and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername), BilledCost MUST adhere to the following requirements:
```

  * The sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MAY differ from the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceDetailId](#datasets.costandusage.invoicedetailid), [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid), and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername) when [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Open".

>   * Otherwise, the sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MUST match the sum of [CostAndUsage.BilledCost](#datasets.costandusage.billedcost) for the same [CostAndUsage.InvoiceDetailId](#datasets.costandusage.invoicedetailid), [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid), and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername).

Author: In Feb 24 TF-1, @pydi-aws raised the good point that the typical step-down in precision between a billing dataset (where cost may be specified out to, say, 6 digits) and an issued invoice (where cost must be specified according to the smallest portion of the billing currency, which for USD is 2 digits) results in a natural and recurring discrepancy, owing to rounding errors.  For a customer at scale, this value may be, say, $50 or $100.  The acceptable discrepancy from one aggregation slice to the next is likely different for every consumer, depending on the scale of their consumption.

For FOCUS, that means that `Cost and Usage` and `Invoice Detail` will never tie exactly, other than for the most trivial examples.

The question becomes: is there a way for us to add this expectation of acceptable, trivial rounding errors between `Cost and Usage` and `Invoice Detail`, while still preserving the static nature of these conformance requirements?

@pydi-aws and I will think about this, and we will come back to Mar 3 TF-1 to discuss further.

Author: @pydi-aws 
I have now implemented a change to handle for expected rounding errors while retaining the static nature of the requirement.  Essentially, for the comparison, the discrepancy must be smaller than:

> Tolerance = ([CostAndUsage](#datsets.costandusage) rows) x 0.5 x (Smallest Subunit of [BillingCurrency](#datasets.invoicedetail.billingcurrency))

See the newly-added Implementation Guidance section, which runs down a couple of examples.

I'm pretty happy with this!  It keeps the check static, and it works with a dataset of any size.  The only drawback is that any validator must have the cardinality of the dataset on-hand in order to perform the check.  I think that's a reasonable expectation, but I'll confer with @mike-finopsorg and TF-RM to see if they have any concerns.

Let me know if you have any feedback and/or if we can resolve!  I will bring this to Mar 3 TF-1 to cover off.  Thanks for bringing this up -- super important catch!

Reviewer 2: Which file is the "Implementation Guidance section" under? searching for Implementation Guidance I only find your comment in this PR.

Author: @mike-finopsorg 
The definition for this column (i.e., `BilledCost`) -- but this thread is against an outdated commit.

See [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files/#diff-4feedf3b866f48dbf233870b74bcdf95527641af7ac681bdae58365d1d447808).

Author: @mike-finopsorg and I huddled to discuss.  His feedback was that, while the static check is the right direction, the tolerance was far too wide for solutions at scale.

I went back to the drawing board and have made a couple of changes:

* Am now using the _square root_ of row count, rather than row count itself.
* Implement a floor of 100 * the smallest subunit of currency.

Please see the latest definition of BilledCost [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files#diff-4feedf3b866f48dbf233870b74bcdf95527641af7ac681bdae58365d1d447808).

This should now be much closer to the reality we expect for datasets of all sizes.  Thanks for pushing on this, Mike!

Author: Resolving this in favor of [this non-outdated comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files#r2872646903).

Billed Cost is denominated in the [Billing Currency](#datasets.invoicedetail.billingcurrency). Billed Cost is commonly used to support FinOps activities, including invoice reconciliation, [*cash-flow-based*](#glossary:cash-based-accounting) forecasting, budgeting, and cost allocation.

## Requirements

TODO: Apply the results of the work from FR #982.

BilledCost adheres to the following requirements:

* BilledCost MUST be of type Decimal.
* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be 0 for *charges* where payments are received by a third party (e.g., marketplace transactions).
* BilledCost MUST be denominated in the BillingCurrency.

> * The sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same ID generated by the [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername).

Reviewer 4: In the Invoice Handling attribute (PR #1501), the following normative requirements have been identified as being specific to this column.

> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MUST match the payable amount provided in the corresponding invoice line items when InvoiceDetail.InvoiceStatus is "Closed".
> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MUST match the sum of CostAndUsage.BilledCost for the same CostAndUsage.InvoiceDetailId, CostAndUsage.InvoiceId, and CostAndUsage.InvoiceIssuerName when InvoiceDetail.InvoiceStatus is "Closed".
> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MAY differ from the sum of CostAndUsage.BilledCost for the same CostAndUsage.InvoiceDetailId, CostAndUsage.InvoiceId, and CostAndUsage.InvoiceIssuerName when InvoiceDetail.InvoiceStatus is "Open".

Wording aside, both PR #1501 and this PR both address reconciliation of the sum of BilledCost across the InvoiceDetail and CostAndUsage datasets.

However, there are differences in scope and conditional logic:

* PR #1501 focuses exclusively on InvoiceDetailId and introduces InvoiceStatus as a governing condition (i.e., MUST match when the invoice is Closed, MAY differ when Open), allowing records to temporarily be out of sync while the invoice status is Open.
* This PR addresses both InvoiceId and InvoiceDetailId and permits differences based solely on ChargeCategory = "Tax", without referencing invoice status.

---

**Note:** 

* We have not yet made a final decision on whether dataset-specific and column-specific normative requirements should be specified within attributes that address a particular business context (e.g., Invoice Handling, Discount Handling, etc.), or whether attributes should contain only normative requirements that apply across more than one FOCUS dataset and/or more than one column. (At the time of writing, the referenced requirement is also listed under the Invoice Handling normative requirements.)

Author: @ijurica Understood.  Let's meet to decide the right wording to use here.

Author: @ijurica I have now issued a commit incorporating the Billed Cost requirements as described above.  Please review and resolve this thread if it looks good to you!

Reviewer 4: Looks good. Resolving this thread

* The sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid) MUST match the sum of the payable amount provided in the corresponding invoice.
* When comparing the sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) across the [InvoiceDetail](#datasets.invoicedetail) and [CostAndUsage](#datasets.costandusage) datasets, BilledCost adheres to the following requirements:

>   * The sum of BilledCost for a given [InvoiceId](#datasets.invoicedetail.invoiceid) in [InvoiceDetail](#datasets.invoicedetail) MAY differ from the sum of [BilledCost](#datasets.costandusage.billedcost) that same [InvoiceId](#datasets.costandusage.invoiceid) in [CostAndUsage](#datasets.costandusage) when [ChargeCategory](#datasets.invoicedetail.chargecategory) = "Tax".

Reviewer 5: I agree with the Tax call out. Does it also make sense to make this distinction for Credit or Adjustment charge categories too or is it Tax and non-tax as the behavior distinctions?

Author: Tax is the only exception/carvout that I believe we should allow.  Everything else should already be in Cost and Usage.  If it's not: why not?  

Lmk if you think we can resolve.

Reviewer 4: It seems to me that when InvoiceId and InvoiceDetailId is used as the composite key for reconciliation, differences should not be allowed, even for ChargeCategory = "Tax".

Author: Ah.  Yes, I think you are right, @ijurica.  That's a miss on my end.  InvoiceId should carry a Tax carveout.  InvoiceDetailId should not, as we are requiring the inclusion of ChargeCategory, which specifies Tax.

I think this normative requirement on line 24 here should remain (though we need to move to the Dataset.Column format you have proposed in #1501), but line 26 needs to shift.  The thing is that we still need to specify that a given InvoiceDetailId may be in Invoice Detail, but may be missing in Cost and Usage.

Reviewer 5: Agreed with @ijurica's correction. The Tax carveout makes sense at the 
InvoiceId level, but not at InvoiceDetailId since ChargeCategory is already 
included at that grain.

@shawnalpay your updated position on line 24 vs. 26 tracks for me. The 
InvoiceDetailId requirement needs revision to remove the Tax exception.

Author: I have updated this verbiage per @ijurica's suggestion.  @Matt-Cowsert please review and resolve.

Reviewer 5: Looks good to me, thanks @shawnalpay for applying it and @ijurica for the recommendation. Resolving
